### PR TITLE
The Queue constructor should take an initial set of objects

### PR DESCRIPTION
**Pull request by chrisseaton**
Sorry I broke https://github.com/ruby/ruby/pull/3768 while trying to divide it up. GitHub won't let me re-open it!

Implements https://bugs.ruby-lang.org/issues/17327, including feedback from @ko1, @nobu, @baweaver.

`Set` specs moved to https://github.com/ruby/spec/pull/830.

### CODE REVIEW

File: thread_sync.c [hunk 1]
```diff
@@ -839,15 +839,29 @@ queue_closed_result(VALUE self, struct rb_queue *q)
...
 {
+    VALUE initial;
     struct rb_queue *q = queue_ptr(self);
     RB_OBJ_WRITE(self, &q->que, ary_buf_new());
     list_head_init(queue_waitq(q));
+    rb_scan_args(argc, argv, "01", &initial);
+    if (argc == 1) {
+        RB_OBJ_WRITE(self, &q->que, rb_convert_type(initial, T_ARRAY, "Array", "to_ary"));
+    } else {
+        RB_OBJ_WRITE(self, &q->que, ary_buf_new());
```
Comment (nobu) [inline thread_sync.c:863]: else block seems unnecessary

Comment (chrisseaton) [reply thread_sync.c:862]: Gah sorry - actually it's the previous assignment of `q->que` that was redundant. I've pushed. What we're doing is setting the queue's array in the first place to a `dup` of the array from the `Enumerable`, or a new empty array.

Comment (nobu) [reply thread_sync.c:862]: `ary_buf_new` creates a hidden array.
I think concatenation to it is safer than dup.

Comment (chrisseaton) [reply thread_sync.c]: Thanks done.

**Pull request merged**